### PR TITLE
Add /device/stats endpoint

### PR DESCRIPTION
**Pull request by robiscoding**
# Description

- Allow users using a docker container to get the current usage stats snapshot of their inference container

## Type of change

Please delete options that are not relevant.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [X] New feature (non-breaking change which adds functionality)
-   [ ] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

- Tested running cpu container locally and hitting /device/stats endpoint
- Example response:
```json
{
"status": 200,
"message": "done",
"container_stats": {
...
"num_procs": 0,
"storage_stats": {},
"cpu_stats": {
"cpu_usage": {
"total_usage": 3582820000,
...
},
"system_cpu_usage": 1078509950000000,
...
}
```

## Any specific deployment considerations

For example, documentation changes, usability, usage/costs, secrets, etc.

## Docs

-   [ ] Docs updated? What were the changes:

### CODE REVIEW

**Comment by robiscoding** [comment]
I added a test but it ended up just being mostly mocks because of the subprocess calls so didn't seem useful, I can add back if we should have it

File: inference/core/utils/container.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        return socket.gethostname()
+    except Exception as e:
+        logger.warn(f"Error getting container ID: {e}")
+        return None
+
+
+def get_container_stats():
+    container_id = get_container_id()
+    if container_id:
+        result = subprocess.run(
```
Comment (PawelPeczek-Roboflow) [inline inference/core/utils/container.py:19]: for that to work you need to have docker socket exported into the container? Is that ensured somehow?

File: inference/core/utils/container.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+                "--unix-socket",
+                "/var/run/docker.sock",
+                f"http://localhost/containers/{container_id}/stats?stream=false",
+            ],
+            capture_output=True,
+            text=True,
+        )
+        try:
+            return json.loads(result.stdout)
+        except json.JSONDecodeError:
```
Comment (PawelPeczek-Roboflow) [inline inference/core/utils/container.py:31]: I am against silent error passing

File: inference/core/utils/container.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import socket
+import subprocess
+
+from inference.core.logger import logger
+
+
+def get_container_id():
+    try:
+        return socket.gethostname()
+    except Exception as e:
```
Comment (PawelPeczek-Roboflow) [inline inference/core/utils/container.py:11]: returning node does not make much of a value here

**Review by PawelPeczek-Roboflow — COMMENTED** [review]
I believe the feature which is added is responsibility of top-level manager of inference container, not the container itself. Why do we implement that this way?

**Comment by PawelPeczek-Roboflow** [comment]
Closing in favour of https://github.com/roboflow/inference/pull/712

**Pull request closed without merging**